### ISSUE TITLE
New FusedAdam (since #424 commit 8 Aug 2019) memory Footprint

### ISSUE THREAD

**Issue by vince62s** [NVIDIA/apex#475]
Team,

This huge PR #424 was not squashed and commits in a long period of time trigger lots of side effects.

As an example, if I use FusedAdam as of this commit 4a8c4ac then the memory footprint in GPUs enables to train with a certain batch size (eg 3072 tokens per batch).

On master, I have to reduce significantly the batch size not to get a CUDA OOM.
(use case is OpenNMT-py)

Not easy to give you numbers, but maybe with some test cases you can better reproduce this issue.

**Comment by FDecaYed** [comment]
In terms of accuracy, updated fusedadam incoporated this upstream fix
https://github.com/pytorch/pytorch/commit/fed5ca192c98619947cca2fb9491ac9624b787d2
Result will be slightly different compare to before, but should be correct and same as upstream adamW now. Do you observe any regression in term of final accuracy? This fix does affect initial steps more.

In terms of memory
updated fused adam no longer fuses grad unscaling and update, thus now fp32 grad is instantiated. This could be the main source of difference if you are comparing against old fusedadam.

There could also be other change affect how memory get reused, but those should not be strict more use. With the dynamic nature of pytorch's allocator, any change on order and life cycle of tensor could cause OOM if you are very close to limit, even though you are not using more memory in total.

**Comment by vince62s** [comment]
I'll go for a full run and check if there is some regression, thanks for the explanation.


> In terms of memory
> updated fused adam no longer fuses grad unscaling and update, thus now fp32 grad is instantiated. This could be the main source of difference if you are comparing against old fusedadam.

What is the rationale for this change ?

**Comment by FDecaYed** [comment]
We want fused optimizers to work with AMP O1, while not doing it by special casing in both optimizer and amp backend.
We expect speed benefit to be marginal(in your case ~1%) and memory benefit also small(most of the case optimizer is not peak of caching allocator). So trading it for expandibility and maintainability is seem a good deal.

@mcarilli speaking of optimizer not being peak. I believe on 4a8c4ac FusedAdam is still handled by FP16_optimizer path. Which means fp32 grad never exist and fp16 grad become None after zero_grad()
In latest master, only fp32 becomes None after zero_grad(), fp16 grad will be zeroed but keep existing through forward/backward(higher potential causing OOM compare to before)
any thought on changing this as well to improve that?
https://github.com/NVIDIA/apex/blob/master/apex/amp/_process_optimizer.py#L366

**Comment by vince62s** [comment]
OK, I trained a bit longer but I am pretty sure something is not so good.
Here is training of base Tranformer for the first 15k steps (on 4 Gpu)

Old FusedAdam:
[2019-09-05 14:58:39,291 INFO] Step 14000/50000; acc:  87.11; ppl:  1.56; xent: 0.44; lr: 0.00075; 99195/112956 tok/s;   8319 sec
[2019-09-05 15:08:35,610 INFO] Step 15000/50000; acc:  87.16; ppl:  1.55; xent: 0.44; lr: 0.00072; 98597/112762 tok/s;   8915 sec
[2019-09-05 15:08:35,709 INFO] number of examples: 3000
[2019-09-05 15:08:43,581 INFO] Validation perplexity: 11.5924
[2019-09-05 15:08:43,581 INFO] Validation accuracy: 62.038

New FusedAdam:
[2019-09-06 11:14:15,308 INFO] Step 14000/50000; acc:  83.72; ppl:  1.83; xent: 0.60; lr: 0.00075; 99332/114016 tok/s;   8272 sec
[2019-09-06 11:24:08,907 INFO] Step 15000/50000; acc:  83.71; ppl:  1.83; xent: 0.60; lr: 0.00072; 99135/113397 tok/s;   8865 sec
[2019-09-06 11:24:09,016 INFO] number of examples: 3000
[2019-09-06 11:24:16,982 INFO] Validation perplexity: 11.0455
[2019-09-06 11:24:16,982 INFO] Validation accuracy: 60.5374

Losing 1 to 2 BLEU depending on the testset.

**Comment by FDecaYed** [comment]
Hi @vince62s 
Have you got any final result shows poor convergence? 
Also, the best thing we can try is probably test your code with latest upstream AdamW. That is our 'golden standard' here. If both of that and new fusedadam shows worse convergence, then it will be very interesting to investigate whether the fix causes this and start a wider conversation.

I have tested the new fused adam against latest upstream pytorch adamW, with this example.
https://pytorch.org/tutorials/beginner/pytorch_with_examples.html#pytorch-optim
And new fusedadam is a lot closer to AdamW on fp32. On amp O2, I don't see any meaningful difference either, everything seems to be within numerical error range.

**Comment by vince62s** [comment]
My O2 example above had already converged.
I'll try AdamW but as a matter of fact there is a difference between old and new fusedadam, not only because of the "minor fix" you mentioned. Or it is not minor.

**Comment by FDecaYed** [comment]
How big is the difference on final result?

It is 'minor' per upstream. [https://github.com/pytorch/pytorch/pull/22628](url)
And I definitely agree with you. We need to find out whether the change you're seeing is from this fix(meaning the fix is not minor), or there is other issue on our side.

**Comment by vince62s** [comment]
@mcarilli @FDecaYed 

It seems it is not related to the "minor" fix at all.
Let me explain.

At OpenNMT-py, before the integration of the new Apex API, (this commit: https://github.com/OpenNMT/OpenNMT-py/commit/aaa220b753c9749ac1a0d86a32e4c1fdcb4faa7a)
we used to test both Adam, FusedAdam and it gave similar results (hence, with FP16_Optimizer)
Let me call this "Good results"

After the above commit, Adam in FP16 was using the new API when FusedAdam was still using the FP16_Optimizer path as mentioned by @FDecaYed in a previous post.

Since then we did all our training with FusedAdam.

Starting Aug 7/8 with the merge of #424 FusedAdam switched to the new API.
We started to have weird results (reason why I opened this issue).

We have tested several run with the same preprocessed data, and all hyperparameters being the same. Here is what we have:
Old API (as stated above): Good results for both FusedAdam and Adam
NewAPI: Wrong results for both FusedAdam and Adam
FP32 Adam without Apex: Good results
pytorch 1.1 or 1.2 has no impact (hence 'minor fix' no impact)

Bottom line: the issue is with the new API or the way we use the new API.
However using the New API with the old path for FusedAdam was fine, so we think our calls to the new API are ok.

We really need to find out what is going on because if we want no regression we would have to go back to FP16_Optimizer.

NB: I can give you some detailed numbers by email if needed. Also The dataset can be downloaded and I can give you a small script so that you can replicate.

Cheers.
Vincent

**Comment by mcarilli** [comment]
https://github.com/OpenNMT/OpenNMT-py/blob/97ad4c66997d6d4d80290f66ddc3f5dab6557ab5/onmt/utils/optimizers.py#L89-L90

Looks to me like your integration isn't right.  You're passing a list with 2 models, but only receiving one on the RHS.  Also, if `model.generator` is a submodule of `model`, you don't need to pass both.  I would try changing this to
```
        model, optimizer = apex.amp.initialize(
            model,
            optimizer,
            opt_level=opt.apex_opt_level,
            loss_scale=loss_scale,
            keep_batchnorm_fp32=False if opt.optim == "fusedadam" else None)
```

**Comment by vince62s** [comment]
Your suggestion does not work (we tested it as a first intention at the beg, and I retried it) it spits a cast error.
I just added a list on the LHS and getting the same results.

Not even talking about FusedAdam, is there anywhere a regression test between old API (FP16_optimizer) and New API, with plain Adam, on a real word dataset (not just a toy set) ?

**Comment by FDecaYed** [comment]
Does amp api with pytorch adam work before #424 ?
if not
then problem is amp(or how you use it), not new fusedadam. you happend to start see this now because old fusedadam fall back to fp16_optimizer and hide the problem.
if it works, then only difference between that and now should be the optimizer.

This question will help us on where we look into next.

Some other points:
- both pytorch 1.1/2 have old adam/adamW. If we want to test effect of the optimizer fix, we need 1.3 adamW.(we may rule out this without testing since above question)
- it seems grad clipping code in your repo is breaking for fp16, no matter what. This should not be the reason of this issue though.

**Comment by vince62s** [comment]
To your first question:
it works in the sense it runs, but results are not good, significantly worse than before amp.
So yes, the problem is either amp or the way we use it.

However, when we use amp the way we use it, with old fusedadam then it works, maybe because it falls back to fp16çoptimizer, but at least it shows that we call amp correctly.
right?

**Comment by mcarilli** [comment]
For reference, how does accuracy look with O1?  In general O1 is always best practice, and the new FusedAdam should permit it (ie if that was the only reason you used O2, it is no longer relevant, but maybe there is some other reason).

**Comment by FDecaYed** [comment]
For your way of using amp O2, I feel it is probably ok
you passed in a list of model and model.generator, but never caught the return. So both of them will be cast inplace and your old handle to them should now see fp16 model correctly.
I do agree with carilli that O1 is much well maintained. If O1 works, and since it can be use with fusedadam now, then switching to it might be the correct way moving forward.

For clipping:
I think 'update_master_grads' and 'clip_master_grads' are things in old **fp16_utils.fp16_optimizer**. So after you switch to amp api, neither real amp, nor old FusedAdam(fall back to **optimizers.fp16_optimizer**)have those. So clipping code won't be hit ever since then.
Plus, `apex.amp.master_params(self._optimizer)` should be called instead of `apex.amp.master_params(self)`. If this code path is ever executed, you should see 
`'Optimizer' object has no attribute 'param_groups'` already

**Comment by vince62s** [comment]
Just finished the first 5000 iterations on 4 GPU with keep_batchnorm_fp32=True ==> no change.
Trying O1 now, but I fear that for our use case not keeping a master copy of weights in fp32 might be an issue.

**Comment by mcarilli** [comment]
With O1, the model weights are left as FP32 (in other words, the model weights are the master weights).  O1 is more conservative about casting to FP32 for internal functions (which is why it's recommended as an out-of-the-box approach) so it may use more memory, but hopefully not much.

**Comment by vince62s** [comment]
it actually starts much much better, letting it run with Adam first, will do Fusedadam just after.
Now, it means something might be wrong with O2 then.

**Comment by mcarilli** [comment]
O2 was our original hypothesized recipe for mixed precision, but maintaining separate master weights is more confusing for everyone.  IMO the only remaining utility of O2 is to support certain internal use cases.  O1 is what I hope all public codes are able to use, and O1 is the implementation that's I'm working on for Amp upstream (https://github.com/pytorch/pytorch/issues/25081).  The API will be different (as requested by upstream) but more flexible and powerful.

**Comment by vince62s** [comment]
one thing though:
O1 Adam 61k tok/sec
O1 FusedAdam 69k tok/sec
O2 FusedAdam is 79k tok/sec
is O1 supposed to be that slower ?

**Comment by mcarilli** [comment]
It depends.  Currently O1 in Apex relies on a lot of Python side patching logic, so if any sections of your model are CPU bound, those sections may slow down.  The upstream integration of O1-style casting will be entirely on the C++ side, and be faster.

**Comment by FDecaYed** [comment]
@mcarilli Using old fusedadam thus clipping by group is also an known issue. We are working on our bert fine-tuning example to see what's the best way to address that.
@vince62s The only known different from pytorch adamW(make sure it is the adamW and post 1.2) and new fusedadam is the fix. We would really want to rule that out first. 
You can try that(easist way is probably copy paste TOT adamw.py, since that's a python only change in upstream) or you can provide me a repro so I can help. A dockerfile or versions would be nice to rule out other differences.

**Comment by vince62s** [comment]
@FDecaYed  I am not sure what we want to prove here.
Based on my previous post, trying to clarify:
1) New FusedAdam with AMP/O1 looks unstable but does not mean it is algorithmically wrong. Do you want me to compare with AdamW with AMP/O1 ?
2) New FusedAdam without AMP does not work. Or maybe I am not getting the proper way to use it without AMP.

Hope this is clearer.
Anyhow I'll give a try to AdamW when I get time.

**Comment by FDecaYed** [comment]
1. I'm trying to rule out whether the unstable is caused by the algo change. Thus comparing new fusedadam against new AdamW, both with O1 should tell.
if o1 with new AdamW does not work, then the upstream change is the problem. if it works, then new fusedadam has its own problem.
2. New FusedAdam without AMP does not work, meaning you have tried it with fp32 no amp?

**Comment by vince62s** [comment]
ok...
1. Same crash with AdamW + O1
2. I misunderstood the comment in FusedAdam. I thought "without amp" meant "old FP16_optimizer path". So yes New FusedAdam with FP32 works fine. What I meant is that with current apex master, there is no way to use FusedAdam with old API, correct?

Anyhow, we now need to find out why AdamW makes things unstable with FP16.

**Comment by FDecaYed** [comment]
If new upstream adamw with O1 also causes the crash(and I suspect upstream adamw with o2 also possibly cause similar accuracy drop we see in new fusedadam with o2), then it does seems the algorithm change causes issue when work with fp16

I feel It’s a problem worth looking into. Let’s probably start with a repro and add back old algo to fused Adam

**Comment by FDecaYed** [comment]
@mcarilli shall we look into and track this? 
Latest results seems indicate new upstream adam/W seems unstable using with amp(not relate to fused though).
